File: scripts/scrape_bytedance_jobs.js
#!/usr/bin/env bun

const fs = require("fs");
const path = require("path");

let chromium;
try {
  ({ chromium } = require("playwright"));
} catch (error) {
  console.error(
    "Missing dependency: playwright. Install it locally or run Bun with NODE_PATH pointing at a node_modules directory that contains Playwright."
  );
  console.error(String(error));
  process.exit(1);
}

const USER_AGENT =
  "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/135.0.0.0 Safari/537.36";
const DEFAULT_LIST_URL =
  "https://jobs.bytedance.com/campus/position?keywords=&category=6704215862557018372&location=&project=7194661126919358757&type=&job_hot_flag=&current=1&limit=10&functionCategory=&tag=";
const DEFAULT_CHROME_PATH = "/usr/bin/google-chrome";
const DEFAULT_PAGE_SIZE = 600;

const CSV_COLUMNS = [
  "source",
  "company",
  "position_id",
  "parent_position_id",
  "job_requirement_id",
  "position_intention_id",
  "position_intention_name",
  "position_name",
  "position_url",
  "position_req_code",
  "batch_id",
  "batch_name",
  "category_name",
  "family_code",
  "family_name",
  "data_source",
  "work_locations",
  "interview_locations",
  "departments",
  "circles",
  "circle_codes",
  "channels",
  "feature_tags",
  "publish_time",
  "modify_time",
  "graduation_from",
  "graduation_to",
  "requirement",
  "description",
];

function parseArgs(argv) {
  const args = {
    outputDir: "data",
    listUrl: DEFAULT_LIST_URL,
    projectId: "7194661126919358757",
    projectSlug: "byteintern",
    projectName: "ByteIntern",
    categoryId: "6704215862557018372",
    categorySlug: "backend",
    categoryName: "后端",
    company: "ByteDance",
    source: "bytedance",
    pageSize: DEFAULT_PAGE_SIZE,
    chromePath: process.env.BYTEDANCE_CHROME_PATH || DEFAULT_CHROME_PATH,
    headless: true,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (!token.startsWith("--")) continue;
    const key = token.slice(2);
    if (key === "no-headless") {
      args.headless = false;
      continue;
    }
    const value = argv[index + 1];
    if (value === undefined) {
      throw new Error(`Missing value for argument ${token}`);
    }
    index += 1;
    switch (key) {
      case "output-dir":
        args.outputDir = value;
        break;
      case "list-url":
        args.listUrl = value;
        break;
      case "project-id":
        args.projectId = value;
        break;
      case "project-slug":
        args.projectSlug = value;
        break;
      case "project-name":
        args.projectName = value;
        break;
      case "category-id":
        args.categoryId = value;
        break;
      case "category-slug":
        args.categorySlug = value;
        break;
      case "category-name":
        args.categoryName = value;
        break;
      case "page-size":
        args.pageSize = Number.parseInt(value, 10);
        break;
      case "chrome-path":
        args.chromePath = value;
        break;
      default:
        throw new Error(`Unknown argument: ${token}`);
    }
  }

  if (!Number.isFinite(args.pageSize) || args.pageSize <= 0) {
    throw new Error(`Invalid page size: ${args.pageSize}`);
  }

  return args;
}

function buildPositionUrl(positionId) {
  return `https://jobs.bytedance.com/campus/position/${positionId}/detail`;
}

function joinValues(values) {
  if (!values) return "";
  if (Array.isArray(values)) {
    return values
      .map((value) => String(value ?? "").trim())
      .filter(Boolean)
      .join(" | ");
  }
  return String(values);
}

function toPublishTime(value) {
  if (!value) return "";
  if (typeof value === "number") {
    return new Date(value).toISOString();
  }
  return String(value);
}

function normalizeJob(job, args) {
  const projectName =
    job?.job_subject?.name?.zh_cn ||
    job?.job_subject?.name?.i18n ||
    job?.job_subject?.name?.en_us ||
    args.projectName;
  const categoryName = job?.job_category?.name || args.categoryName;
  const workLocations = joinValues(
    (job?.city_list || []).map((city) => city?.name || city?.i18n_name || city?.en_name).filter(Boolean)
  );
  const departmentId = job?.department_id ? String(job.department_id) : "";
  const featureTags = [];
  if (job?.job_hot_flag) featureTags.push("hot");
  if (job?.process_type != null) featureTags.push(`process_type:${job.process_type}`);
  if (job?.storefront_mode != null) featureTags.push(`storefront_mode:${job.storefront_mode}`);

  return {
    source: args.source,
    company: args.company,
    position_id: String(job?.id || ""),
    parent_position_id: "",
    job_requirement_id: "",
    position_intention_id: "",
    position_intention_name: "",
    position_name: job?.title || "",
    position_url: buildPositionUrl(job?.id || ""),
    position_req_code: job?.code || "",
    batch_id: job?.job_subject?.id || args.projectId,
    batch_name: projectName,
    category_name: categoryName,
    family_code: "",
    family_name: "",
    data_source: "",
    work_locations:
      workLocations ||
      job?.city_info?.name ||
      job?.city_info?.i18n_name ||
      job?.city_info?.en_name ||
      "",
    interview_locations: "",
    departments: departmentId,
    circles: "",
    circle_codes: "",
    channels: "campus",
    feature_tags: joinValues(featureTags),
    publish_time: toPublishTime(job?.publish_time),
    modify_time: "",
    graduation_from: "",
    graduation_to: "",
    requirement: job?.requirement || "",
    description: job?.description || "",
  };
}

function writeJson(filePath, payload) {
  fs.writeFileSync(filePath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
}

function escapeCsvCell(value) {
  const text = String(value ?? "");
  if (text.includes('"') || text.includes(",") || text.includes("\n") || text.includes("\r")) {
    return `"${text.replaceAll('"', '""')}"`;
  }
  return text;
}

function writeCsv(filePath, rows) {
  const lines = [CSV_COLUMNS.join(",")];
  for (const row of rows) {
    lines.push(CSV_COLUMNS.map((column) => escapeCsvCell(row[column] || "")).join(","));
  }
  fs.writeFileSync(filePath, `${lines.join("\n")}\n`, "utf8");
}

async function bootstrapAndFetchAll(args) {
  const browser = await chromium.launch({
    headless: args.headless,
    executablePath: args.chromePath,
  });
  const page = await browser.newPage({ userAgent: USER_AGENT });

  try {
    await page.goto(args.listUrl, { waitUntil: "domcontentloaded", timeout: 120000 });
    await page.waitForFunction(
      () => document.cookie.includes("atsx-csrf-token="),
      {},
      { timeout: 30000 }
    );

    return await page.evaluate(async ({ categoryId, projectId, pageSize }) => {
      const token = decodeURIComponent(
        document.cookie
          .split("; ")
          .find((entry) => entry.startsWith("atsx-csrf-token="))
          ?.split("=")[1] || ""
      );

      async function fetchPage(limit, offset) {
        const payload = {
          keyword: "",
          limit,
          offset,
          job_category_id_list: [categoryId],
          tag_id_list: [],
          location_code_list: [],
          subject_id_list: [projectId],
          recruitment_id_list: [],
          portal_type: 3,
          job_function_id_list: [],
          storefront_id_list: [],
          portal_entrance: 1,
        };

        const response = await fetch("/api/v1/search/job/posts", {
          method: "POST",
          credentials: "include",
          headers: {
            accept: "application/json, text/plain, */*",
            "content-type": "application/json",
            "portal-channel": "campus",
            "portal-platform": "pc",
            "website-path": "campus",
            "x-csrf-token": token,
            env: "undefined",
          },
          body: JSON.stringify(payload),
        });

        if (!response.ok) {
          const text = await response.text();
          throw new Error(`ByteDance API failed: ${response.status} ${text.slice(0, 500)}`);
        }

        const body = await response.json();
        return body.data;
      }

      const firstPage = await fetchPage(pageSize, 0);
      const total = Number(firstPage.count || 0);
      const jobs = [...(firstPage.job_post_list || [])];

      for (let offset = jobs.length; offset < total; offset += pageSize) {
        const nextPage = await fetchPage(pageSize, offset);
        jobs.push(...(nextPage.job_post_list || []));
      }

      return {
        total,
        cookies: document.cookie,
        jobs,
      };
    }, { categoryId: args.categoryId, projectId: args.projectId, pageSize: args.pageSize });
  } finally {
    await browser.close();
  }
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const result = await bootstrapAndFetchAll(args);
  const rows = result.jobs.map((job) => normalizeJob(job, args));
  const generatedAt = new Date().toISOString();

  const outputDir = path.resolve(process.cwd(), args.outputDir);
  fs.mkdirSync(outputDir, { recursive: true });

  const fileStem = `${args.source}_positions_${args.projectSlug}_${args.categorySlug}`;
  const jsonPath = path.join(outputDir, `${fileStem}.json`);
  const csvPath = path.join(outputDir, `${fileStem}.csv`);

  writeJson(jsonPath, {
    generated_at: generatedAt,
    source: args.source,
    company: args.company,
    project_id: args.projectId,
    project_name: args.projectName,
    category_id: args.categoryId,
    category_name: args.categoryName,
    total_count: result.total,
    jobs: rows,
  });
  writeCsv(csvPath, rows);

  console.log(`ByteDance positions: ${result.total}`);
  console.log(`Wrote ${jsonPath}`);
  console.log(`Wrote ${csvPath}`);
}

main().catch((error) => {
  console.error(error instanceof Error ? error.stack || error.message : String(error));
  process.exit(1);
});

File: scripts/scrape_bytedance_jobs.py
from __future__ import annotations

import argparse
from datetime import datetime, timezone
from pathlib import Path

try:
    from scripts.scrape_campus_jobs import USER_AGENT, write_csv, write_json
except ModuleNotFoundError:  # pragma: no cover - direct script execution path
    from scrape_campus_jobs import USER_AGENT, write_csv, write_json

try:
    from playwright.sync_api import sync_playwright
except ImportError as exc:  # pragma: no cover - environment dependent
    raise SystemExit(
        "Missing dependency: playwright for Python. Install it before running this script."
    ) from exc


DEFAULT_LIST_URL = (
    "https://jobs.bytedance.com/campus/position?keywords=&category=6704215862557018372"
    "&location=&project=7194661126919358757&type=&job_hot_flag=&current=1&limit=10"
    "&functionCategory=&tag="
)
DEFAULT_CHROME_PATH = "/usr/bin/google-chrome"
DEFAULT_PAGE_SIZE = 600


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Scrape ByteDance campus job listings into JSON and CSV.")
    parser.add_argument("--output-dir", default="data")
    parser.add_argument("--list-url", default=DEFAULT_LIST_URL)
    parser.add_argument("--project-id", default="7194661126919358757")
    parser.add_argument("--project-slug", default="byteintern")
    parser.add_argument("--project-name", default="ByteIntern")
    parser.add_argument("--category-id", default="6704215862557018372")
    parser.add_argument("--category-slug", default="backend")
    parser.add_argument("--category-name", default="后端")
    parser.add_argument("--company", default="ByteDance")
    parser.add_argument("--source", default="bytedance")
    parser.add_argument("--page-size", type=int, default=DEFAULT_PAGE_SIZE)
    parser.add_argument("--chrome-path", default=DEFAULT_CHROME_PATH)
    parser.add_argument("--no-headless", action="store_true")
    return parser.parse_args()


def build_position_url(position_id: str) -> str:
    return f"https://jobs.bytedance.com/campus/position/{position_id}/detail"


def join_values(values: list[str]) -> str:
    return " | ".join(value for value in values if value)


def to_publish_time(value: object) -> str:
    if value is None:
        return ""
    if isinstance(value, (int, float)):
        return datetime.fromtimestamp(value / 1000, tz=timezone.utc).isoformat()
    return str(value)


def normalize_job(job: dict[str, object], args: argparse.Namespace) -> dict[str, str]:
    project_name = (
        ((job.get("job_subject") or {}).get("name") or {}).get("zh_cn")
        or ((job.get("job_subject") or {}).get("name") or {}).get("i18n")
        or ((job.get("job_subject") or {}).get("name") or {}).get("en_us")
        or args.project_name
    )
    category_name = ((job.get("job_category") or {}).get("name")) or args.category_name
    cities = job.get("city_list") or []
    work_locations = join_values(
        [
            city.get("name") or city.get("i18n_name") or city.get("en_name") or ""
            for city in cities
            if isinstance(city, dict)
        ]
    )
    feature_tags: list[str] = []
    if job.get("job_hot_flag"):
        feature_tags.append("hot")
    if job.get("process_type") is not None:
        feature_tags.append(f"process_type:{job['process_type']}")
    if job.get("storefront_mode") is not None:
        feature_tags.append(f"storefront_mode:{job['storefront_mode']}")

    return {
        "source": args.source,
        "company": args.company,
        "position_id": str(job.get("id") or ""),
        "parent_position_id": "",
        "job_requirement_id": "",
        "position_intention_id": "",
        "position_intention_name": "",
        "position_name": str(job.get("title") or ""),
        "position_url": build_position_url(str(job.get("id") or "")),
        "position_req_code": str(job.get("code") or ""),
        "batch_id": str(((job.get("job_subject") or {}).get("id")) or args.project_id),
        "batch_name": str(project_name),
        "category_name": str(category_name),
        "family_code": "",
        "family_name": "",
        "data_source": "",
        "work_locations": work_locations
        or str(((job.get("city_info") or {}).get("name")) or ((job.get("city_info") or {}).get("i18n_name")) or ""),
        "interview_locations": "",
        "departments": str(job.get("department_id") or ""),
        "circles": "",
        "circle_codes": "",
        "channels": "campus",
        "feature_tags": join_values(feature_tags),
        "publish_time": to_publish_time(job.get("publish_time")),
        "modify_time": "",
        "graduation_from": "",
        "graduation_to": "",
        "requirement": str(job.get("requirement") or ""),
        "description": str(job.get("description") or ""),
    }


def fetch_all_jobs(args: argparse.Namespace) -> dict[str, object]:
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(
            headless=not args.no_headless,
            executable_path=args.chrome_path,
        )
        page = browser.new_page(user_agent=USER_AGENT)
        try:
            page.goto(args.list_url, wait_until="domcontentloaded", timeout=120000)
            page.wait_for_function(
                "() => document.cookie.includes('atsx-csrf-token=')",
                timeout=30000,
            )
            return page.evaluate(
                """
                async ({ categoryId, projectId, pageSize }) => {
                  const token = decodeURIComponent(
                    document.cookie
                      .split('; ')
                      .find((entry) => entry.startsWith('atsx-csrf-token='))
                      ?.split('=')[1] || ''
                  );

                  async function fetchPage(limit, offset) {
                    const payload = {
                      keyword: '',
                      limit,
                      offset,
                      job_category_id_list: [categoryId],
                      tag_id_list: [],
                      location_code_list: [],
                      subject_id_list: [projectId],
                      recruitment_id_list: [],
                      portal_type: 3,
                      job_function_id_list: [],
                      storefront_id_list: [],
                      portal_entrance: 1,
                    };

                    const response = await fetch('/api/v1/search/job/posts', {
                      method: 'POST',
                      credentials: 'include',
                      headers: {
                        accept: 'application/json, text/plain, */*',
                        'content-type': 'application/json',
                        'portal-channel': 'campus',
                        'portal-platform': 'pc',
                        'website-path': 'campus',
                        'x-csrf-token': token,
                        env: 'undefined',
                      },
                      body: JSON.stringify(payload),
                    });

                    if (!response.ok) {
                      const text = await response.text();
                      throw new Error(`ByteDance API failed: ${response.status} ${text.slice(0, 500)}`);
                    }

                    return response.json();
                  }

                  const first = await fetchPage(pageSize, 0);
                  const firstData = first.data || {};
                  const total = Number(firstData.count || 0);
                  const jobs = [...(firstData.job_post_list || [])];

                  for (let offset = jobs.length; offset < total; offset += pageSize) {
                    const next = await fetchPage(pageSize, offset);
                    jobs.push(...((next.data || {}).job_post_list || []));
                  }

                  return {
                    total,
                    token,
                    cookies: document.cookie,
                    jobs,
                  };
                }
                """,
                {
                    "categoryId": args.category_id,
                    "projectId": args.project_id,
                    "pageSize": args.page_size,
                },
            )
        finally:
            browser.close()


def main() -> None:
    args = parse_args()
    result = fetch_all_jobs(args)
    rows = [normalize_job(job, args) for job in result["jobs"]]
    generated_at = datetime.now(timezone.utc).isoformat()

    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    file_stem = f"{args.source}_positions_{args.project_slug}_{args.category_slug}"
    json_path = output_dir / f"{file_stem}.json"
    csv_path = output_dir / f"{file_stem}.csv"

    write_json(
        json_path,
        {
            "generated_at": generated_at,
            "source": args.source,
            "company": args.company,
            "project_id": args.project_id,
            "project_name": args.project_name,
            "category_id": args.category_id,
            "category_name": args.category_name,
            "total_count": result["total"],
            "jobs": rows,
        },
    )
    write_csv(csv_path, rows)

    print(f"ByteDance positions: {result['total']}")
    print(f"Wrote {json_path}")
    print(f"Wrote {csv_path}")


if __name__ == "__main__":
    main()
